frontend/apps/reader/modules/readerlink.lua

```diff
@@ -687,137 +775,30 @@ function ReaderLink:onGotoLink(link, neglect_current_location, allow_footnote_po
-        callback = function()
-            UIManager:close(dialog)
-            UIManager:show(QRMessage:new{
-                text = link_url,
-                width = Device.screen:getWidth(),
-                height = Device.screen:getHeight()
-            })
-        end,
-    })
-    if self.ui.wallabag then
-        table.insert(alt_handlers_buttons, {
-            text = _("Add to Wallabag"),
-            callback = function()
-                UIManager:close(dialog)
-                self.ui:handleEvent(Event:new("AddWallabagArticle", link_url))
-            end,
-        })
-    end
-    if Device:canOpenLink() then
-        table.insert(alt_handlers_buttons, {
-            text = _("Open in browser"),
-            callback = function()
-                UIManager:close(dialog)
-                Device:openLink(link_url)
-            end,
-        })
-    end
-
-    -- Check if it is a wikipedia link
-    local wiki_lang, wiki_page = link_url:match([[https?://([^%.]+).wikipedia.org/wiki/([^/]+)]])
-    if wiki_lang and wiki_page then
-        logger.dbg("Wikipedia link:", wiki_lang, wiki_page)
-        -- Ask for user confirmation before launching lookup (on a
-        -- wikipedia page saved as epub, full of wikipedia links, it's
-        -- too easy to click on links when wanting to change page...)
-        -- But first check if this wikipedia article has been saved as EPUB
-        local epub_filename = wiki_page .. "."..string.upper(wiki_lang)..".epub"
-        local epub_fullpath
-        -- either in current book directory
-        local last_file = G_reader_settings:readSetting("lastfile")
-        if last_file then
-            local current_book_dir = last_file:match("(.*)/")
-            local safe_filename = util.getSafeFilename(epub_filename, current_book_dir):gsub("_", " ")
-            local epub_path = current_book_dir .. "/" .. safe_filename
-            if util.pathExists(epub_path) then
-                epub_fullpath = epub_path
-            end
-        end
-        -- or in wikipedia save directory
-        if not epub_fullpath then
-            local dir = G_reader_settings:readSetting("wikipedia_save_dir")
-            if not dir then dir = G_reader_settings:readSetting("home_dir") end
-            if not dir then dir = require("apps/filemanager/filemanagerutil").getDefaultDir() end
-            if dir then
-                local safe_filename = util.getSafeFilename(epub_filename, dir):gsub("_", " ")
-                local epub_path = dir .. "/" .. safe_filename
-                if util.pathExists(epub_path) then
-                    epub_fullpath = epub_path
+    local default_title =  T(_("External link:\n\n%1"), BD.url(link_url))
+    local title = default_title
+
+    for _, fn_button in pairs(self._external_link_buttons) do
+        local button = fn_button(self, link_url)
+        if button.show_in_dialog_func then
+            local show, button_title = button.show_in_dialog_func()
```

Comment:
You will be calling each button.show_in_dialog_func() twice:
- once here, to know if it provides a title, and build the final `title`
- once below with self:getButtonsForExternalLinkDialog
```lua
    self.external_link_dialog = ButtonDialogTitle:new{
        title = title,
        buttons = self:getButtonsForExternalLinkDialog(link_url),
    }
```
May be your self:getButtonsForExternalLinkDialog() should itself build and return the title? 
`local buttons, title = self:getButtonsForExternalLinkDialog(link_url)`

Reply:
Excellent (: ty